packages/material-ui-lab/src/NProgressBar/nProgressState.js

```diff
@@ -0,0 +1,80 @@
+import * as React from 'react';
+
+function create(createState) {
```

Comment:
Why do we need this over standard `React.useSate`? We should start with standard state managment and then discuss more advanced solutions. State managment is not a trivial matter especially in concurrent React so I'd rather not slide in a custom implementation when introducing a new component.

Reply:
> Why do we need this over standard `React.useSate`? 

This is a global state so we can keep the API outside of the component. I agree that we could use `useState` hook, but then the app would to be wrappeed by the component itselt. So we would need to create 3 new things:

- A Context Provider
- A Context Consumer
- A Context Hook

In my opinion, this seemed way more work than having a rather simple global state implementation. And the API feels much better with a global state since we're talking about a "global progress".

> We should start with standard state managment and then discuss more advanced solutions.

Sorry for skipping to the second part, but I already went throught a lot of thought and research about standard state management, and in the previous topic I mentioned my opinion about, so now we can discuss about it. I'm also expecting some thought from @oliviertassinari on https://github.com/mui-org/material-ui/pull/25889#issuecomment-825404153.

> State managment is not a trivial matter especially in concurrent React so I'd rather not slide in a custom implementation when introducing a new component.

I totally agree. Concurrent React is complex but for complex state. The progress (and state changes) will always happen outside of this context and I currently don't see any drawbacks but if the component is approved as is, the point of having it on Lab would be exactly for testing if this would work on all use cases before going core.

--

These are my opinion and represent as far as my knowledge goes. If after this read you still feel the same way, I will gladly change the code myself if you have any suggestion.

Reply:
> In my opinion, this seemed way more work than having a rather simple global state implementation.

That seems fine to me considering the global state implementation is not simple to begin with.

> And the API feels much better with a global state since we're talking about a "global progress".

That is an important limitation that I don't see documented or guarded against in any way. It sounds like there can only be a single `NProgress` component on a document?

> the point of having it on Lab would be exactly for testing if this would work on all use cases before going core.

Note that concurrent mode testing is not that trivial at the moment and dog-feeding is quite limited at the moment. But this implementation does have known problems with regard to "tearing". That's why `useSubscription` exists.

> Concurrent React is complex but for complex state

Concurrent React is not complex per se. But it does have certain rules that apply to everything regardless of its complexity.

Reply:
> That seems fine to me considering the global state implementation is not simple to begin with.

Agreed.

> That is an important limitation that I don't see documented or guarded against in any way. It sounds like there can only be a single `NProgress` component on a document?

Yes, that's exactly it. The component is positioned with `fixed` and on the top by default. If you add more than one, they'll overlap, so there's no actual need for having separate state.

> Note that concurrent mode testing is not that trivial at the moment and dog-feeding is quite limited at the moment. But this implementation does have known problems with regard to "tearing". That's why `useSubscription` exists.

No argument here, this actually goes further than my knowledge.

> Concurrent React is not complex per se. But it does have certain rules that apply to everything regardless of its complexity.

Yes, you are totally right. I also didn't express myself in the way I actually meant. The point is that on what we're working right now, global state, we only keep one value that will only exist and be used on a specific situation where the concurrent mode doesn't apply. The state change can only be triggered when the component is already loaded and rendered, and he doesn't trigger rerender on no other component than itself.